### PR TITLE
:book:  Add blog post for Fake Node Scalability on Metal3

### PR DESCRIPTION
**Pull request by mquhuy**
This blog post explains the process we had in deploying a Scaled-up Metal3 workflow on 1000 single-node clusters, using only fake nodes. The process takes into use Fake IPA and Fake Kubernetes API Server (FKAS).

### CODE REVIEW

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (Rozzii) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:45]:
```suggestion
In a normal Metal3 workflow, BMO translates the kubernetes reconciling logic to concrete actions and forward them to Ironic, which as part of the provisioning process tells the IPA what to do. Once the BMH is provisioned, Ironic will only monitor the power state of the machine and IPA will stop running right before the BMH switches to "provisioned" state, CAPI can then set up Kubernetes clusters on these provisioned machines.
```

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (Rozzii) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:52]: "special networks", not really, we use separate provisioning network but it is not a hard requirement.

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (Rozzii) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:61]: Missing link for Helm, or the link syntax is not needed.

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (Rozzii) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:65]: Other missing  link

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (Rozzii) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:70]: on the host machine instead of in the host machine

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (Rozzii) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:83]: missing link for the script

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (Rozzii) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:190]: I would suggest getting rid of languge like "we won't , we will just ..."   It is just my personal opinion but usually in technical writings this personal tone is improper. But I know in blog posts even in technical topics folks use it. 

Up to you, I would not use it.

**Review by Rozzii — CHANGES_REQUESTED** [review]
Generally a very interesting blog post, and it is the result of a lot of work done by both you and Mohamed so thank you for that !

I have a few comments, nothing major.

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (mquhuy) [reply _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:190]: Thanks Adam. That's a good advice, the reason I used a rather casual language was to make this post a bit friendlier to newcomers (Same reason why I went on and on about explaining how Metal3, ironic, etc. work), but yeah, I agree that those words aren't always needed.

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (mquhuy) [reply _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:52]: I think ironic requires `baremetal` network, doesn't it?

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (mquhuy) [reply _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:61]: Yeah. As this a WIP, I was writing down markdown link syntax just as placeholders. No proper links were added. Probably I could add the TODO to remember adding them later.

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (Rozzii) [reply _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:52]: e.g in a scenario where we use refish with virtual media booting and there is no network seapration between the NICs including the BMC NIC we can use 1 network.

**Comment by tuminoid** [comment]
I'll review later when it is no longer WIP and it is passing basic checks.

**Review by lentzi90 — COMMENTED** [review]
I'm not sure what kind of review is requested here since this is still WIP and is failing multiple tests. Do you want a review of the structure or the content/idea?

Since I don't know what is expected I have just skimmed a little and will comment on a couple of things that the tests may be missing.

1. Please check previous posts for how the "[frontmatter](https://jekyllrb.com/docs/front-matter/)" should be used (the first few lines of the file).
2. In general I think it will be easier to "write properly" from the start instead of hunting down all little typos, missing links etc. If you feel the flow and just want to write, fine. Do that but don't commit until you have gone through it at least once more. I also find it helpful to keep a TODO list at the end of the file with things I am aware about or have planned but didn't do yet. This helps the reviewer know if it is worth commenting or not when reviewing WIP content.
3. There is a mix of ticks in this post. Please check carefully if you use `'` or `’`. I think we should be using `'` everywhere.

**Comment by kashifest** [comment]
Apologies for the confusion, lets wait until the PR is ready for review.

**Comment by mquhuy** [comment]
> I'm not sure what kind of review is requested here since this is still WIP and is failing multiple tests. Do you want a review of the structure or the content/idea?
> 
> Since I don't know what is expected I have just skimmed a little and will comment on a couple of things that the tests may be missing.
> 
> 1. Please check previous posts for how the "[frontmatter](https://jekyllrb.com/docs/front-matter/)" should be used (the first few lines of the file).
> 2. In general I think it will be easier to "write properly" from the start instead of hunting down all little typos, missing links etc. If you feel the flow and just want to write, fine. Do that but don't commit until you have gone through it at least once more. I also find it helpful to keep a TODO list at the end of the file with things I am aware about or have planned but didn't do yet. This helps the reviewer know if it is worth commenting or not when reviewing WIP content.
> 3. There is a mix of ticks in this post. Please check carefully if you use `'` or `’`. I think we should be using `'` everywhere.

Thanks Lennart. In this stage I would appreciate if you could give general comments on how do you feel about the content, what to include, or what to drop to make the post more appeal, etc.

When the post is no longer a WIP, I think its structure will be in a more concrete state, at which it's a bit more difficult to change.

**Comment by mquhuy** [comment]
/hold 
Need to merge https://github.com/Nordix/metal3-clusterapi-docs/pull/88 and reference everything properly in the blog post.

**Review by tuminoid — COMMENTED** [review]
Wrap the lines at 80. It is not enforced yet, but let's not add more, as we should fix it eventually.

**Comment by mquhuy** [comment]
> Wrap the lines at 80. It is not enforced yet, but let's not add more, as we should fix it eventually.

Thank you. Done now.

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (lentzi90) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:9]: I suggest dropping this heading. It is technically directly below the title, so it is two headings on top of each other. That is not a big deal but it also makes the excerpt look strange since there it isn't visible as a heading. It looks like this:

> Introduction If you’ve ever tried scaling out Kubernetes clusters in a bare-metal environment, you’ll know that large-scale testing comes with serious challenges. Most of us don’t have access to enough physical servers—or even virtual machines—to simulate the kinds of large-scale environments we need for stress testing, especially when deploying hundreds...

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (lentzi90) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:106]:
```suggestion
[cloud-init](https://cloud-init.io/) will install Kubernetes components on the
```
I would personally avoid putting so much details here because many of them may not always be correct. Cloud-init *can* be used, but so can ignition, cloud-init *may* install kubernetes components, or they may be already in the image, etc...

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (lentzi90) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:170]:
```suggestion
`PROVISIONING_IP`, hence the need to have a specialized `configMap.`
```

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (lentzi90) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:175]: This looks quite strange rendered when one part of the word is in ticks and the other outside. I would just skip the ticks. They aren't usually used for product names anyway.
```suggestion
MariaDB-ironic database) stays in another pod. A list of PROVISIONING_IPs is
```

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (lentzi90) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:184]:
```suggestion
an init Container for ironic, and its job is to download the IPA image
to a Persistent Volume. This image contains the __Ironic Python
```

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (lentzi90) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:199]:
```suggestion
[here](https://github.com/Nordix/metal3-clusterapi-docs/tree/main/Support/Multitenancy/Scalability-with-Fake-Nodes/ironic).
```

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (lentzi90) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:261]:
```suggestion
```

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (lentzi90) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:249]: Remove the blank line and consider breaking up the long line.
```suggestion
mkdir cert
kubectl get secret -n baremetal-operator-system \
  ironic-cert -o json -o=jsonpath="{.data.ca\.crt}" \
  | base64 -d >cert/ironic-ca.crt
```

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (lentzi90) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:270]: A comment explaining this would be good

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (lentzi90) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:406]:
```suggestion
Note that in this experiment, we provisioned every one of the 1000 fake nodes to a
```

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (lentzi90) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:467]:
```suggestion
[here](https://github.com/Nordix/metal3-clusterapi-docs/blob/main/Support/Multitenancy/Scalability-with-Fake-Nodes/README.md).
```

**Review by lentzi90 — COMMENTED** [review]
Great job @mquhuy !
Just some small comments below.

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (lentzi90) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:272]: Sorry but this is not correct. It is an htpasswd file and the username is `admin` (first part of the string). The rest is probably a bcrypt hash of the password, which I don't know what it is. Anyway you can probably just mention that this an htpasswd file used for securing the container(s)? I'm not sure if it is used for sushy-tools or fake IPA or both though.

**Review by lentzi90 — COMMENTED** [review]
/approve

File: _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md
Comment (Rozzii) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:104]:
```suggestion
installed on the nodes by cloud-init, and once the process succeeds, Ironic passes control
```

Comment (Rozzii) [inline _posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md:104]: Leave out this ironic passing control and stuff. 

Ironic doesn't do any sort of handover, it doesn't even know about K8s or cloud-init.  Instead mention that after reboot the IPA is no longer running and Ironic only have power control and BMC access nothing else.

**Review by Rozzii — CHANGES_REQUESTED** [review]
Looks much better, thanks for the fixes, I still have a little bit of issue how the provisioning is presented here.

**Comment by Rozzii** [comment]
/hold
just until my change request is fulfilled.

**Comment by lentzi90** [comment]
Comments have been addressed so I'm adding lgtm back
/lgtm

**Pull request merged**
